### ISSUE TITLE
Add ability to reorder dataset files as draggable elements

### ISSUE THREAD

**Issue by tjanek** [IQSS/dataverse#5280]
On unpublished dataset, it would be nice to add a feature for reordering dataset files using drag and drop and saving that information for eg. in database.
New feature (draggable elements) should be available only while dataset is unpublished on such pages:
 * when uploading new files.
 * editing files on "Files" tab (new context menu element in "Edit Files", for eg. "Reorder").

**Comment by djbrooke** [comment]
Hi @tjanek -we unfortunately won't be able to provide design guidance on this to allow implementation in the near term. Our design efforts are focused in other areas right now and we don't have the time to do the research, user testing, and design that would be needed for this change. Note that there's similar functionality desired in #4959. Let me know if you'd be still interested in contributing to the project and we're happy to point to some issues that are a bit more ready to be picked up.

**Comment by tjanek** [comment]
Hi @djbrooke, we want to have our own instance of Dataverse with some extra functionalities that are missing in current version, for eg. possibility of reordering files from UI and some others like new domain specific metadata blocks, etc.
#4959 describes files order functionality as a something that is configured at installation step of the application and we would like to have this as UI feature for our users.

**Comment by pdurbin** [comment]
I'm passing along something from MrK at http://irclog.iq.harvard.edu/dataverse/2018-11-08#i_78357

> "Hi, my friend responded to the yesterdays issue https://github.com/IQSS/dataverse/issues/5280 , and i just wanted to add that in the future, we will be able to merge those features into your instance if you would wish so."

I noticed that there's already a branch at https://github.com/IQSS/dataverse/compare/develop...tjanek:5280-dataset-files-reordering !

**Comment by djbrooke** [comment]
Hi @tjanek - when we have the capacity to explore the reordering from a design perspective, I hope that we can take advantage of the code that you've written. It's just not something we can code review and bring into the develop branch right now. 

It's great to hear about the work towards the new Dataverse instance. When you're in production, feel free to email us at support@dataverse.org and we'll put you on the map on the project site.

**Comment by tjanek** [comment]
@pdurbin Yes, there is a branch on our fork for this feature, we trying to implement reordering dataset files right now but also we treat this task as to get some familiar with Dataverse codebase.

@djbrooke OK, I understand that you don't have time to review code right now :) Also, this feature propably take us some time to implement as we just getting started with the codebase :)
Work towards our Dataverse instance will be a long-term process (3-4 months to finish).

**Comment by pdurbin** [comment]
@tjanek 3-4 months is long enough that I'm hoping we can merge some pull requests for you. Fixing bugs is always less controversial so if you find any you want to work on (or want suggestions of which ones to work on), please just let us know! Thanks!

**Comment by pdurbin** [comment]
@tjanek I was reminded today by @djbrooke that @qqmyers listed "Custom Sorting" in his "QDR and Dataverse Coordination" doc. I'm not sure if @qqmyers would want me to link to the doc here or not but I don't think he's object to me copy/pasting the relevant sections:

* Custom Sorting
   * https://github.com/IQSS/dataverse/issues/4959 
   * Not necessarily sorting, but grouping - goal is to privilege the documentation files for users
   * Consider limiting to out of the box tags
   * Needs some design checks - seems like a solution for a specific case, but need to know how it plays in with other dataset changes in progress
   * One potential design - a setting that is an ordered list of some tags, initial ordering of files is sorted by these tags, alphabetical within each tag, files without any listed tag alphabetical at the end. Off/fully alphabetical by default.
* Allow for custom sorting of files – for us, in particular, we want to sort documentation files to the top of a data project and are currently clumsily doing so by adding 0_ to the filename        
   * Ex. https://data.qdr.syr.edu/dataset.xhtml?persistentId=doi:10.5064/F6BUAX58
   * Existing user request (for Dataverses) at https://github.com/IQSS/dataverse/issues/3066
   * You might want to look at how a paginator is being added to the file list. There is a screenshot at https://github.com/IQSS/dataverse/issues/4656#issuecomment-405655071  and (currently) a live demo at https://dev1.dataverse.org/dataset.xhtml?persistentId=doi:10.5072/FK2/WGWEOM
    * Next Steps
      * Danny - Create Github issue focused on installation level setting - alphanumeric, date, tag, etc.

In short, it looks like #4959 was created to capture the desire for this feature but I thought you might appreciate the additional detail above. And perhaps you and @qqmyers can talk.

In the context of a meeting on file hierarchy today another suggestion for organizing files is by directory. Maybe there's be a directory at the top called "000" or something. Still clumsy, though. Please see also https://github.com/IQSS/dataverse/issues/3448#issuecomment-444633647

**Comment by djbrooke** [comment]
Hey @tjanek, we've implemented the updated sort options in #5584, which was released in 4.15. I'm not in favor of implementing this drag and drop in the application as it doesn't scale well to larger datasets and I think maintaining this level of dataset-by-dataset customization will be a challenge in the future. I'm going to close this for now, but if there's more community interest for this level of sort customization, others should feel free to re-open this.
